src/service/rate_limit.rs

```diff
@@ -0,0 +1,105 @@
+use crate::envoy::{RateLimitDescriptor, RateLimitRequest};
+use crate::service::Service;
+use protobuf::{Message, RepeatedField};
+use proxy_wasm::hostcalls::dispatch_grpc_call;
+use proxy_wasm::types::Status;
+use std::time::Duration;
+
+const RATELIMIT_SERVICE_NAME: &str = "envoy.service.ratelimit.v3.RateLimitService";
+const RATELIMIT_METHOD_NAME: &str = "ShouldRateLimit";
+pub struct RateLimitService<'a> {
+    endpoint: String,
+    metadata: Vec<(&'a str, &'a [u8])>,
+}
+
+impl<'a> RateLimitService<'a> {
+    pub fn new(endpoint: &str, metadata: Vec<(&'a str, &'a [u8])>) -> RateLimitService<'a> {
+        Self {
+            endpoint: String::from(endpoint),
+            metadata,
+        }
+    }
+    pub fn message(
+        domain: String,
+        descriptors: RepeatedField<RateLimitDescriptor>,
+    ) -> RateLimitRequest {
+        RateLimitRequest {
+            domain,
+            descriptors,
+            hits_addend: 1,
+            unknown_fields: Default::default(),
+            cached_size: Default::default(),
+        }
+    }
+}
+
+fn grpc_call(
+    upstream_name: &str,
+    initial_metadata: Vec<(&str, &[u8])>,
+    message: RateLimitRequest,
+) -> Result<u32, Status> {
+    let msg = Message::write_to_bytes(&message).unwrap(); // TODO(didierofrivia): Error Handling
+    dispatch_grpc_call(
+        upstream_name,
+        RATELIMIT_SERVICE_NAME,
+        RATELIMIT_METHOD_NAME,
+        initial_metadata,
```

Comment:
I guess this should be `self.metadata` and `initial_metadata` disappears from that method signature, right?

Reply:
I've made the `grpc_call` outside the `impl` block so we can mock it in the tests, that's why it doesn't take a `self` param. Should it be include it within the `RateLimitService` impl?